homeassistant/components/manual_mqtt/alarm_control_panel.py

```diff
@@ -410,34 +447,48 @@ def _validate_code(self, code, state):
         check = not alarm_code or code == alarm_code
         if not check:
             _LOGGER.warning("Invalid code given for %s", state)
+            mqtt.async_publish(
+                self.hass, self._status_topic, self._payload_invalid, self._qos, False
+            )
+
         return check
 
     @property
     def device_state_attributes(self):
         """Return the state attributes."""
-        if self.state != STATE_ALARM_PENDING:
+        if self.state != STATE_ALARM_PENDING and self.state != STATE_ALARM_ARMING:
             return {}
         return {
-            ATTR_PRE_PENDING_STATE: self._previous_state,
-            ATTR_POST_PENDING_STATE: self._state,
+            ATTR_PREVIOUS_STATE: self._previous_state,
+            ATTR_NEXT_STATE: self._state,
         }
 
     async def async_added_to_hass(self):
-        """Subscribe to MQTT events."""
+        """Run when entity about to be added to hass & subscribe to MQTT Events."""
+        await super().async_added_to_hass()
+
         async_track_state_change_event(
             self.hass, [self.entity_id], self._async_state_changed_listener
         )
 
         async def message_received(msg):
             """Run when new MQTT message has been received."""
-            if msg.payload == self._payload_disarm:
-                await self.async_alarm_disarm(self._code)
-            elif msg.payload == self._payload_arm_home:
-                await self.async_alarm_arm_home(self._code)
-            elif msg.payload == self._payload_arm_away:
-                await self.async_alarm_arm_away(self._code)
-            elif msg.payload == self._payload_arm_night:
-                await self.async_alarm_arm_night(self._code)
+            try:
+                payload = json.loads(msg.payload)
```

Comment:
We don't want to accept json. That's too complicated.

Reply:
> We don't want to accept json. That's too complicated.

Then how would you suggest we obtain an MQTT command with a requested state change and a code from the panel, so that HA can perform the code verification?  Right now the code might as well not exist as far as MQTT is concerned!  I'm happy to split this into a different PR (as I just said).

Reply:
@derekatkins I agree with you that it is a step in the right direction to allow users to send a code together with an arm/disarm command over MQTT.
But if you want to add this feature, please do so in a generic way.
Your change requires adapting all systems that currently (or in the future) use this integration., to the choices that you made here.
To my opinion, that doesn't fit with a component that is intended for generic usage.

Suggestion: take a look at the [MQTT light](https://www.home-assistant.io/integrations/light.mqtt/) integration, which deals with similar behaviour.
This integration allows to send a command for turning on a light, but also permits the user to define a brightness, color, effect etc.
JSON is certainly not the only option here.
If you want to add support for JSON payloads, please allows users to define their own `json_attributes_template`. If you need an example, refer to the MQTT light again.

Reply:
@nielsfaber Thank you for the reference, I will look in those places.  I do want to point out that I made the JSON payload optional, so existing users of this particular interface do not need to make any changes and the implementation will continue to work as it's been working forever (insecurely, mind you).  The change, as implemented, allows the panel to optionally send in a code, by using JSON.  We picked JSON because it seemed the most lightweight option and would allow expansion down the road.

Today is a family day so I don't know how much time I'll have to go research MQTT light, but I will do so ASAP.  Thanks!